Typecheck python

tpoterba opened the pull request:
I haven't ported type checking to everything yet -- just hail context. It'll take an hour or two to write it for everything, and I want to wait until after breakingbad is merged to do that to avoid a hellish rebase (also in case you want me to change the interface).

cseed commented on python/hail/typecheck.py:
```diff
@@ -0,0 +1,123 @@
+        for i, arg in enumerate(args):
+            argname = argspec[i]
+            tc = kw.get(argname)
+            if tc:
```
You're allowing typecheck to be optional?  Any reason for that?

cseed commented on python/hail/context.py:
```diff
@@ -44,6 +47,16 @@ class HailContext(object):
+               master=nullable(strlike),
+               local=strlike,
+               log=strlike,
+               quiet=only(bool),
```
I think only shouldn't be necessary here.  Actually, you set `tc = only(tc)` below, so just bool should work.

cseed replied:
Ah, you use straight bool below.

cseed commented on python/hail/context.py:
```diff
@@ -44,6 +47,16 @@ class HailContext(object):
     :vartype sc: :class:`.pyspark.SparkContext`
     """
 
+    @typecheck(sc=nullable(SparkContext),
```
The interface is perfect.  Great work.

cseed commented on python/hail/typecheck.py:
```diff
@@ -0,0 +1,123 @@
+        for kwarg in kw.keys():
+            if not kwarg in argspec:
+                raise RuntimeError("invalid typecheck signature: unmatched parameter '%s'" % kwarg)
+        for i, arg in enumerate(args):
```
It doesn't look like your are type checking the keyword arguments here.  Only the  positional arguments appear in `args`:

```
>>> def f(*args, **kwargs):
...   return (args, kwargs)
... 
>>> f(1, 2, k=5)
((1, 2), {'k': 5})
```

cseed commented on python/hail/typecheck.py:
```diff
@@ -0,0 +1,123 @@
+
+class TypeChecker(object):
+    def __init__(self, f, supported):
+        self.f = f
```
I think this would be clearer if you had an interface with two functions: check (your f) and pretty/toString builds the message recursively.

danking replied:
For `pretty`, I would add a `nested` argument. That way, when things show up at the top-level there is no unnecessary parentheses. E.g.

```
dict of Any, Any

dict of Any, (dict of Any, Any)
```

cseed commented on python/hail/typecheck.py:
```diff
@@ -0,0 +1,123 @@
+from decorator import decorator, getargspec
+from types import ClassType
+import re
```
This should be in its own package defined in __init__ and declare exports so people who import * from typecheck don't get extract and DictChecker and other things.

tpoterba replied:
I made it a package and wrote a separate set of tests for it

tpoterba commented:
Addressed comments. 
 - Refactored to a separate module and added module-level tests. 
 - Cleaned up TypeChecker interface to call recursively down

tpoterba commented:
I'm not particularly happy about having to check explicitly for a "self" argument. While it's python convention to name the thing self, it's not required.

cseed commented on python/hail/typecheck/check.py:
```diff
@@ -0,0 +1,202 @@
+
+    def _typecheck(f, *args, **kwargs):
+        argspec = getargspec(f)
+        argspec_args = argspec.args
```
It this necessary?

cseed commented on python/hail/typecheck/check.py:
```diff
@@ -0,0 +1,202 @@
+        self_cls = None
+
+        # strip the 'self' parameter, since this does not need to be checked
+        if len(argspec_args) > 0 and argspec_args[0] == 'self':
```
self is not a special name, it is convention.  Can you get the scope where f is defined, and check if it is a class?

tpoterba replied:
no, I don't think so. I spent three hours trying to do this yesterday, but an instance method is just a method. I am a bit stuck

cseed commented on python/hail/typecheck/check.py:
```diff
@@ -0,0 +1,202 @@
+        if len(argspec_args) > 0 and argspec_args[0] == 'self':
+            # try to get more information about the class to write better error messages
+            try:
+                self_cls = args[0].__class__
```
You realize this might be wrong if the user's call is in error (which you're trying to check?)

tpoterba replied:
This can only be wrong in one of two cases:

1. somebody defines a method that's NOT an instance method with a self parameter:

```
def f(self, x, y):
    return x + y

f(1,2,3)
```

2. Somebody uses an instance method using the expanded syntax, without passing in the correct object:

```
class Foo:
    def bar(self, x, y):
        return 5 + x + y

Foo.bar(1,2,3)
```

cseed commented on python/hail/typecheck/check.py:
```diff
@@ -0,0 +1,202 @@
+            check_spec = argspec_args
+
+        # name of function, preceded by the class name if the _f_ is an instance method
+        name = '%s%s()' % (extract(self_cls).split('.')[-1] + '.' if self_cls else '', f.__name__)
```
Also, f might be defined on a superclass of the class of the first argument.  So you'll get B.f, but f is defined on A for B extends A.

cseed commented on python/hail/typecheck/check.py:
```diff
@@ -0,0 +1,202 @@
+
+
+def typecheck(*targs, **kw):
+    if len(targs) != 0 and len(kw) != 0:
```
Seems reasonable to only support keyword args.  That's all you use, anyway.

cseed commented on python/hail/typecheck/check.py:
```diff
@@ -0,0 +1,202 @@
+        # name of function, preceded by the class name if the _f_ is an instance method
+        name = '%s%s()' % (extract(self_cls).split('.')[-1] + '.' if self_cls else '', f.__name__)
+
+        # ensure that the typecheck signature is appropriate and matches the function signature
```
I still think you should be able to typecheck args and kwargs as normal variables.

danking commented on python/hail/java.py:
```diff
@@ -107,13 +108,13 @@ def handle_py4j(func, *args, **kwargs):
     except py4j.protocol.Py4JError as e:
         if e.args[0].startswith('An error occurred while calling'):
             msg = 'An error occurred while calling into JVM, probably due to invalid parameter types.'
+            _, _, tb = sys.exc_info()
```
I don't see this being used?

tpoterba commented:
At this time, I can see no possible way to check that the function is an instance method. This means we either have to:
1. stop enforcing that every parameter is typechecked
2. require people put the self parameter in the typecheck

I don't really see another way.

tpoterba commented:
I'm happy with this initial interface: typecheck and typecheck_method

Back to you!

tpoterba commented:
Addressed comments. Back to you

tpoterba commented:
I added support for varargs and varkwargs.

cseed commented on python/hail/typecheck/check.py:
```diff
@@ -0,0 +1,206 @@
+    def check(self, x):
+        passes = isinstance(x, dict)
+        if passes:
+            for k, v in x.iteritems():
```
Use all here, too.

cseed commented on python/hail/typecheck/check.py:
```diff
@@ -0,0 +1,206 @@
+    spec = getargspec(f)
+    name = f.__name__
+
+    named_argspec = []
```
I'd rename argspec => argname or something.  ArgSpec is an object associated with a function, looks like spec.args is just a list of string, no "spec" anymore.  Makes the set comparison block below a little surprising.
